internal/store/db/mongo/store.go

```diff
@@ -16,37 +16,190 @@
 package mongo
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db"
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/interfaces"
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/models"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// Client provides a wrapper for Mongo's Client type
+type Client struct {
+	Timeout time.Duration
+	Client  *mongo.Database
+	// unexported Client for Disconnect
+	client *mongo.Client
+}
+
+var mongoCollection = "store"
+
 // Store persists a stored object to the data store.
-func Store(o models.StoredObject) error {
+func (c Client) Store(o models.StoredObject) error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
+	defer cancel()
+
+	_, err := c.Client.Collection(mongoCollection).InsertOne(ctx, &o)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // RetrieveFromStore gets an object from the data store.
-func RetrieveFromStore(appServiceKey string) (objects []models.StoredObject, err error) {
-	return nil, nil
+func (c Client) RetrieveFromStore(appServiceKey string) (objects []models.StoredObject, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
+	defer cancel()
+
+	filter := bson.D{{"appServiceKey", appServiceKey}}
+
+	// find all documents
+	cursor, err := c.Client.Collection(mongoCollection).Find(ctx, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	// iterate through all documents
+	for cursor.Next(ctx) {
+		var p models.StoredObject
+		// decode the document
+		if err = cursor.Decode(&p); err != nil {
+			return nil, err
+		}
+		objects = append(objects, p)
+	}
+
+	// check if the cursor encountered any errors while iterating
+	if err = cursor.Err(); err != nil {
+		return nil, err
+	}
+
+	return objects, nil
 }
 
 // Update replaces the data currently in the store with the provided data.
-func Update(o models.StoredObject) error {
+func (c Client) Update(o models.StoredObject) error {
+	// set filters and updates
+	filter := bson.M{"id": o.ID}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.Timeout)
+	defer cancel()
+
+	// update document
+	updated, err := c.Client.Collection(mongoCollection).ReplaceOne(ctx, filter, &o)
+	if err != nil {
+		return err
+	}
+
+	if updated.ModifiedCount == 0 {
+		return errors.New("no updates performed")
```

Comment:
Mongo throws its own error if it can't find the object. Truthfully, this error was very helpful when debugging and working on nailing down my query. We probably don't need it in production.

What would you propose instead?

Reply:
No sure , but `no updates performed` begs for more information... ;-) Seems not useful to check then if the one case where I would see not updating will be an error above.